File: PythonArchive/Shortest-Peak-Reversal-5m-3x-Margin--main/src/short_trader_multi_filter/optimization_queue.py
from __future__ import annotations

import json
from dataclasses import dataclass, asdict
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, List

from .paths import DATA_DIR


@dataclass
class QueuedOptimization:
    queue_id: str
    queued_at: str
    ready_at: str
    elapsed_seconds: float
    payload: Dict[str, Any]

    def to_dict(self) -> Dict[str, Any]:
        base = asdict(self)
        base["payload"] = self.payload
        return base


class OptimizationQueue:
    def __init__(self, queue_path: Path | str | None = None):
        self.queue_path = Path(queue_path) if queue_path else DATA_DIR / "optimization_queue.json"
        self.queue_path.parent.mkdir(parents=True, exist_ok=True)

    def _load_existing(self) -> List[Dict[str, Any]]:
        if not self.queue_path.exists():
            return []
        try:
            data = json.loads(self.queue_path.read_text())
        except json.JSONDecodeError:
            return []
        return data if isinstance(data, list) else []

    def _persist(self, entries: List[Dict[str, Any]]) -> None:
        self.queue_path.write_text(json.dumps(entries, indent=2))

    def enqueue(self, queued_at: datetime, ready_at: datetime, elapsed_seconds: float, payload: Dict[str, Any]) -> Dict[str, Any]:
        queue_item = QueuedOptimization(
            queue_id=str(int(queued_at.timestamp())),
            queued_at=queued_at.isoformat() + "Z",
            ready_at=ready_at.isoformat() + "Z",
            elapsed_seconds=round(elapsed_seconds, 3),
            payload=payload,
        ).to_dict()

        existing = self._load_existing()
        existing.append(queue_item)
        self._persist(existing)
        return queue_item

File: PythonArchive/Shortest-Peak-Reversal-5m-3x-Margin--main/Unity App Project/short_trader_multi_filter/Assets/Scripts/Core/OptimizationQueue.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;

namespace UnityApp.ShortTraderMultiFilter
{
    public record QueuedOptimization(
        string QueueId,
        string QueuedAt,
        string ReadyAt,
        double ElapsedSeconds,
        Dictionary<string, object> Payload);

    public class OptimizationQueue
    {
        private readonly string _queuePath;

        public OptimizationQueue(string? queuePath = null)
        {
            _queuePath = queuePath ?? Path.Combine(Paths.DataDirectory, "optimization_queue.json");
            Directory.CreateDirectory(Path.GetDirectoryName(_queuePath)!);
        }

        public Dictionary<string, object> Enqueue(DateTime queuedAt, DateTime readyAt, double elapsedSeconds, Dictionary<string, object> payload)
        {
            var queuedOffset = new DateTimeOffset(queuedAt.ToUniversalTime());
            var readyOffset = new DateTimeOffset(readyAt.ToUniversalTime());
            var entry = new QueuedOptimization(
                queuedOffset.ToUnixTimeSeconds().ToString(),
                queuedOffset.ToString("o"),
                readyOffset.ToString("o"),
                Math.Round(elapsedSeconds, 3),
                payload);

            var existing = LoadExisting();
            existing.Add(entry);
            Persist(existing);

            return new Dictionary<string, object>
            {
                ["queue_id"] = entry.QueueId,
                ["queued_at"] = entry.QueuedAt,
                ["ready_at"] = entry.ReadyAt,
                ["elapsed_seconds"] = entry.ElapsedSeconds,
                ["payload"] = entry.Payload
            };
        }

        private List<QueuedOptimization> LoadExisting()
        {
            if (!File.Exists(_queuePath))
            {
                return new List<QueuedOptimization>();
            }

            try
            {
                var content = File.ReadAllText(_queuePath);
                var data = JsonSerializer.Deserialize<List<QueuedOptimization>>(content);
                return data ?? new List<QueuedOptimization>();
            }
            catch
            {
                return new List<QueuedOptimization>();
            }
        }

        private void Persist(List<QueuedOptimization> entries)
        {
            var json = JsonSerializer.Serialize(entries, new JsonSerializerOptions { WriteIndented = true });
            File.WriteAllText(_queuePath, json);
        }
    }
}
